File: getSNODAS_BRB.m
% getSNODAS_BRB.m
% Downloads SNODAS (G02158) data for the Boise River Basin from NSIDC
% using HTTPS (replacing legacy FTP access to sidads.colorado.edu).
%
% Dataset: https://nsidc.org/data/g02158/versions/1
% Resolution: ~1 km (30 arc-sec), daily, 2003-present
% Variables (8 total):
%   1025SlL00 - Precipitation (liquid, kg/m^2)
%   1025SlL01 - Snowfall (solid precip, kg/m^2)
%   1034      - Snow Water Equivalent (m * 1000)
%   1036      - Snow Depth (m * 1000)
%   1038      - Snow pack average temperature (K * 1)
%   1039      - Blowing snow sublimation (m * 100000)
%   1044      - Snow melt (m * 100000)
%   1050      - Snow pack sublimation (m * 100000)
%
% Grid: 6935 x 3351 pixels, flat binary int16, big-endian
% Bounding box: lon [-124.7337, -66.9421], lat [24.9496, 52.8746]
%
% Based on original getSNODAS.m and getSNODASall.m by HP Marshall
% Updated to use HTTPS access, April 2026
%
% Companion scripts:
%   plotSNODAS_BRB.m  - Visualize downloaded data (maps, time series, movies)
%   getSNODAS_WY.m    - Alternative FTP-based downloader (full function)
%   snowViz.m         - Interactive visualization driver
%
% HP Marshall, Boise State University

clear; clc;

%% ====== USER CONFIGURATION ======
% Water year(s) to download
% SNODAS archive begins Oct 2003 (WY2004)
WY = 2021;  % Water year to download (e.g. 2021 = Oct 2020 - Sep 2021)

% Output directory (external data drive - keeps large files out of git repo)
dataRoot = '/Users/hpmarshall/DATA_DRIVE/SnowFusion';
outDir = fullfile(dataRoot, 'SNODAS');
if ~exist(outDir, 'dir'); mkdir(outDir); end

% Temporary directory for extracting tar/gz files
tempDir = fullfile(dataRoot, 'temp_download');
if ~exist(tempDir, 'dir'); mkdir(tempDir); end

%% ====== SNODAS GRID DEFINITION ======
% Full CONUS grid parameters (from SNODAS documentation)
lonmin = -124.733749999999;
lonmax = -66.9420833333342;
latmin = 24.9495833333335;
latmax = 52.8745833333323;
nCols  = 6935;
nRows  = 3351;

lon = linspace(lonmin, lonmax, nCols);
lat = linspace(latmax, latmin, nRows); % north to south

% Bounding box matching UCLA SR tile coverage (latTiles=[43,44], lonTiles=[115,116,117])
% Tiles named by SW corner: W115=[-115,-114], W116=[-116,-115], W117=[-117,-116]
BRB_lat = [43.0 45.0];
BRB_lon = [-117.0 -114.0];

% Find pixel indices for BRB subset
Ix = find(lon >= BRB_lon(1) & lon <= BRB_lon(2));
Iy = find(lat >= BRB_lat(1) & lat <= BRB_lat(2));
lon_sub = lon(Ix);
lat_sub = lat(Iy);

fprintf('BRB subset: %d x %d pixels\n', length(Iy), length(Ix));

%% ====== SNODAS PRODUCT DEFINITIONS ======
% Product codes (used to match filenames inside tar archives)
prodCode = {'1025SlL00', '1025SlL01', '1034', '1036', '1038', '1039', '1044', '1050'};
prodName = {'Precip', 'SnowPrecip', 'SWE', 'Depth', 'Tsnow', 'SublimationBS', 'Melt', 'Sublimation'};
nVars = length(prodCode);

%% ====== BUILD DATE LIST FOR WATER YEAR ======
% Water year runs Oct 1 (year-1) through Sep 30 (year)
startDate = datenum(WY-1, 10, 1);
endDate   = datenum(WY, 9, 30);
allDates  = startDate:endDate;
nDays     = length(allDates);

fprintf('Water Year %d: %s to %s (%d days)\n', WY, ...
    datestr(startDate, 'yyyy-mm-dd'), datestr(endDate, 'yyyy-mm-dd'), nDays);

%% ====== INITIALIZE OUTPUT STRUCTURE ======
Snodas.WY = WY;
Snodas.lat = lat_sub;
Snodas.lon = lon_sub;
Snodas.dates = allDates;
Snodas.datestr = cellstr(datestr(allDates, 'yyyy-mm-dd'));
for v = 1:nVars
    Snodas.(prodName{v}) = NaN(length(Iy), length(Ix), nDays);
end

%% ====== HTTPS BASE URL ======
% NSIDC now serves SNODAS via HTTPS (no authentication required)
% Directory structure: /NOAA/G02158/masked/YYYY/MM_Mon/
baseURL = 'https://noaadata.apps.nsidc.org/NOAA/G02158/masked/';

%% ====== MAIN DOWNLOAD LOOP ======
fprintf('\n=== Downloading SNODAS data for WY%d ===\n', WY);

for d = 1:nDays
    dv = datevec(allDates(d));
    yyyy = dv(1);
    mm   = dv(2);
    dd   = dv(3);

    % Build month directory name (e.g., "01_Jan", "10_Oct")
    monthNames = {'Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'};
    monthDir = sprintf('%02d_%s', mm, monthNames{mm});

    % Build tar filename
    % Convention: SNODAS_YYYYMMDD.tar
    tarName = sprintf('SNODAS_%04d%02d%02d.tar', yyyy, mm, dd);

    % Check if we already have saved .mat file for this date
    matFile = fullfile(outDir, sprintf('SNODAS_BRB_%04d%02d%02d.mat', yyyy, mm, dd));
    if exist(matFile, 'file')
        % Load existing data - check grid size matches current bounding box
        R = load(matFile);
        firstVar = prodName{find(isfield(R, prodName), 1)};
        if ~isempty(firstVar) && ~isequal(size(R.(firstVar)), [length(Iy) length(Ix)])
            fprintf('  [%d/%d] %s - cache size mismatch, re-downloading\n', d, nDays, datestr(allDates(d)));
            delete(matFile);
        else
            for v = 1:nVars
                if isfield(R, prodName{v})
                    Snodas.(prodName{v})(:,:,d) = R.(prodName{v});
                end
            end
            if mod(d, 30) == 0
                fprintf('  [%d/%d] %s - loaded from cache\n', d, nDays, datestr(allDates(d)));
            end
            continue;
        end
    end

    % Build URL
    url = sprintf('%s%04d/%s/%s', baseURL, yyyy, monthDir, tarName);

    fprintf('  [%d/%d] %s ... ', d, nDays, datestr(allDates(d)));

    % Download tar file
    tarFile = fullfile(tempDir, tarName);
    try
        websave(tarFile, url, weboptions('Timeout', 120));
    catch ME
        fprintf('FAILED: %s\n', ME.message);
        continue;
    end

    % Extract tar
    try
        untar(tarFile, tempDir);
        delete(tarFile);
    catch ME
        fprintf('UNTAR FAILED: %s\n', ME.message);
        continue;
    end

    % Process each variable
    dayData = struct();
    for v = 1:nVars
        % Find the .dat.gz file matching this product code
        gzFiles = dir(fullfile(tempDir, ['*' prodCode{v} '*.dat.gz']));
        if isempty(gzFiles)
            continue;
        end

        % Unzip
        gzPath = fullfile(tempDir, gzFiles(1).name);
        gunzip(gzPath);
        delete(gzPath);

        % Find the .dat file
        datFiles = dir(fullfile(tempDir, ['*' prodCode{v} '*.dat']));
        if isempty(datFiles)
            continue;
        end

        % Read binary data: int16, big-endian, 6935 x 3351
        datPath = fullfile(tempDir, datFiles(1).name);
        fid = fopen(datPath, 'r', 'b');  % 'b' = big-endian
        D = fread(fid, [nCols nRows], 'int16');
        fclose(fid);
        D = D';  % transpose to [rows x cols]

        % Subset to BRB
        D_sub = D(Iy, Ix);

        % Store in structure
        Snodas.(prodName{v})(:,:,d) = D_sub;
        dayData.(prodName{v}) = D_sub;

        delete(datPath);
    end

    % Clean up remaining temp files (headers, etc.)
    tempFiles = dir(fullfile(tempDir, '*'));
    for tf = 1:length(tempFiles)
        if ~tempFiles(tf).isdir
            delete(fullfile(tempDir, tempFiles(tf).name));
        end
    end

    % Save individual day .mat file for caching
    dayData.date = allDates(d);
    dayData.lat = lat_sub;
    dayData.lon = lon_sub;
    save(matFile, '-struct', 'dayData');

    fprintf('OK\n');
end

%% ====== SET NO-DATA TO NaN (BEFORE UNIT CONVERSION) ======
fprintf('\nCleaning no-data values...\n');
% SNODAS uses -9999 as fill value in raw int16 data
for v = 1:nVars
    D = Snodas.(prodName{v});
    D(D == -9999) = NaN;
    Snodas.(prodName{v}) = D;
end

%% ====== CONVERT UNITS ======
fprintf('Converting units...\n');

% Precipitation fields: raw values in kg/m^2 scaled by 10
% i.e., value of 100 = 10.0 kg/m^2 = 10.0 mm
Snodas.Precip     = Snodas.Precip / 10;         % -> kg/m^2 (= mm)
Snodas.SnowPrecip = Snodas.SnowPrecip / 10;     % -> kg/m^2 (= mm w.e.)

% SWE and Depth: raw values in meters * 1000
Snodas.SWE   = Snodas.SWE / 1000;    % -> meters
Snodas.Depth = Snodas.Depth / 1000;   % -> meters

% Temperature: raw in Kelvin (no scale factor)
% Snodas.Tsnow stays in Kelvin

% Sublimation and Melt: raw in meters * 100000
Snodas.SublimationBS = Snodas.SublimationBS / 1e5;  % -> meters
Snodas.Sublimation   = Snodas.Sublimation / 1e5;    % -> meters
Snodas.Melt          = Snodas.Melt / 1e5;           % -> meters

%% ====== SAVE COMPLETE WATER YEAR ======
outFile = fullfile(outDir, sprintf('SNODAS_BRB_WY%d.mat', WY));
save(outFile, 'Snodas', '-v7.3');
fprintf('\nSaved: %s\n', outFile);

%% ====== SUMMARY ======
fprintf('\n=== Download Summary ===\n');
fprintf('Water Year: %d\n', WY);
fprintf('Region: Boise River Basin (%d x %d pixels)\n', length(Iy), length(Ix));
fprintf('Variables:\n');
for v = 1:nVars
    nValid = sum(~isnan(Snodas.(prodName{v})(:)));
    nTotal = numel(Snodas.(prodName{v}));
    fprintf('  %s: %.1f%% valid data\n', prodName{v}, 100*nValid/nTotal);
end
% Quick peek: April 1 SWE
apr1 = datenum(WY, 4, 1);
[~, apr1Idx] = min(abs(Snodas.dates - apr1));
apr1SWE = Snodas.SWE(:,:,apr1Idx);
fprintf('\nApril 1 basin statistics:\n');
fprintf('  Mean SWE:  %.1f cm\n', nanmean(apr1SWE(:)) * 100);
fprintf('  Max SWE:   %.1f cm\n', max(apr1SWE(:), [], 'omitnan') * 100);
apr1Depth = Snodas.Depth(:,:,apr1Idx);
fprintf('  Mean Depth: %.1f cm\n', nanmean(apr1Depth(:)) * 100);
fprintf('  Max Depth:  %.1f cm\n', max(apr1Depth(:), [], 'omitnan') * 100);

fprintf('\nOutput file: %s (%.1f MB)\n', outFile, dir(outFile).bytes / 1e6);
fprintf('\nDone! Run plotSNODAS_BRB.m to visualize.\n');

File: python/getSNODAS_BRB.py
"""
getSNODAS_BRB.py
Downloads SNODAS (G02158) data for the Boise River Basin from NSIDC
using HTTPS access.

Dataset: https://nsidc.org/data/g02158/versions/1
Resolution: ~1 km (30 arc-sec), daily, 2003-present

Variables (8 total):
  1025SlL00 - Precipitation (liquid, kg/m^2)
  1025SlL01 - Snowfall (solid precip, kg/m^2)
  1034      - Snow Water Equivalent (m * 1000)
  1036      - Snow Depth (m * 1000)
  1038      - Snow pack average temperature (K * 1)
  1039      - Blowing snow sublimation (m * 100000)
  1044      - Snow melt (m * 100000)
  1050      - Snow pack sublimation (m * 100000)

Grid: 6935 x 3351 pixels, flat binary int16, big-endian
Bounding box: lon [-124.7337, -66.9421], lat [24.9496, 52.8746]

Based on getSNODAS_BRB.m by HP Marshall
Updated to Python, April 2026

HP Marshall, Boise State University
"""

import gzip
import shutil
import tarfile
from datetime import date, timedelta
from pathlib import Path

import numpy as np
import requests

# ====== USER CONFIGURATION ======
WY = 2021  # Water year to download (e.g. 2021 = Oct 2020 - Sep 2021)

DATA_ROOT = Path('/Users/hpmarshall/DATA_DRIVE/SnowFusion')
OUT_DIR   = DATA_ROOT / 'SNODAS'
TEMP_DIR  = DATA_ROOT / 'temp_download'

# ====== SNODAS GRID DEFINITION ======
# Full CONUS grid parameters (from SNODAS documentation)
LON_MIN = -124.733749999999
LON_MAX = -66.9420833333342
LAT_MIN =  24.9495833333335
LAT_MAX =  52.8745833333323
N_COLS  = 6935
N_ROWS  = 3351

# ====== BRB BOUNDING BOX ======
# Matching UCLA SR tile coverage (latTiles=[43,44], lonTiles=[115,116,117])
BRB_LAT = [43.0, 45.0]
BRB_LON = [-117.0, -114.0]

# ====== SNODAS PRODUCT DEFINITIONS ======
PROD_CODES = ['1025SlL00', '1025SlL01', '1034', '1036', '1038', '1039', '1044', '1050']
PROD_NAMES = ['Precip', 'SnowPrecip', 'SWE', 'Depth', 'Tsnow', 'SublimationBS', 'Melt', 'Sublimation']
N_VARS = len(PROD_CODES)

# NSIDC HTTPS base URL (no authentication required)
BASE_URL = 'https://noaadata.apps.nsidc.org/NOAA/G02158/masked/'

MONTH_NAMES = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
               'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']


def build_date_list(wy: int) -> list[date]:
    """Return list of dates for a water year (Oct 1 year-1 through Sep 30 year)."""
    start = date(wy - 1, 10, 1)
    end   = date(wy, 9, 30)
    n = (end - start).days + 1
    return [start + timedelta(days=i) for i in range(n)]


def download_tar(url: str, dest: Path, timeout: int = 120) -> bool:
    """Stream-download a URL to dest. Returns True on success."""
    try:
        with requests.get(url, stream=True, timeout=timeout) as r:
            r.raise_for_status()
            with open(dest, 'wb') as f:
                for chunk in r.iter_content(chunk_size=1 << 16):
                    f.write(chunk)
        return True
    except requests.RequestException as e:
        print(f'FAILED: {e}')
        return False


def extract_tar(tar_path: Path, dest_dir: Path) -> bool:
    """Extract a tar archive to dest_dir. Returns True on success."""
    try:
        with tarfile.open(tar_path) as tf:
            tf.extractall(dest_dir)
        tar_path.unlink()
        return True
    except Exception as e:
        print(f'UNTAR FAILED: {e}')
        return False


def gunzip_file(gz_path: Path) -> Path:
    """Decompress a .gz file in-place, return path to decompressed file."""
    out_path = gz_path.with_suffix('')  # strip .gz
    with gzip.open(gz_path, 'rb') as f_in, open(out_path, 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)
    gz_path.unlink()
    return out_path


def read_snodas_binary(dat_path: Path) -> np.ndarray:
    """
    Read a SNODAS flat binary file.
    Format: int16, big-endian, N_COLS x N_ROWS stored column-major,
    transposed to [N_ROWS x N_COLS] (row = latitude, col = longitude).
    """
    raw = dat_path.read_bytes()
    D = np.frombuffer(raw, dtype='>i2')          # big-endian int16
    D = D.reshape(N_ROWS, N_COLS)                # SNODAS is row-major [rows x cols]
    return D


def load_day_cache(npz_path: Path, expected_shape: tuple) -> dict | None:
    """
    Load a cached daily .npz file.
    Returns the data dict, or None if the file is missing or has a grid size mismatch.
    """
    if not npz_path.exists():
        return None
    data = dict(np.load(npz_path, allow_pickle=True))
    # Check grid dimensions against current bounding box
    first_var = next((n for n in PROD_NAMES if n in data), None)
    if first_var is None:
        return None
    if data[first_var].shape != expected_shape:
        print(f'  cache size mismatch (expected {expected_shape}, '
              f'got {data[first_var].shape}), re-downloading')
        npz_path.unlink()
        return None
    return data


def clean_temp_dir(temp_dir: Path) -> None:
    """Remove all non-directory files from temp_dir."""
    for p in temp_dir.iterdir():
        if p.is_file():
            p.unlink()


def apply_unit_conversions(snodas: dict) -> None:
    """
    Apply SNODAS unit conversions in-place (fill value -9999 -> NaN already applied).
    Precip/SnowPrecip: raw / 10  -> mm (kg/m^2)
    SWE/Depth:         raw / 1000 -> m
    Tsnow:             no change (K)
    Sublimation/Melt:  raw / 1e5  -> m
    """
    snodas['Precip']       = snodas['Precip']       / 10.0
    snodas['SnowPrecip']   = snodas['SnowPrecip']   / 10.0
    snodas['SWE']          = snodas['SWE']           / 1000.0
    snodas['Depth']        = snodas['Depth']         / 1000.0
    # Tsnow: no scaling (already K)
    snodas['SublimationBS'] = snodas['SublimationBS'] / 1e5
    snodas['Melt']         = snodas['Melt']          / 1e5
    snodas['Sublimation']  = snodas['Sublimation']   / 1e5


def main():
    # ---- Create output directories ----
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    TEMP_DIR.mkdir(parents=True, exist_ok=True)

    # ---- Build SNODAS coordinate arrays ----
    lon = np.linspace(LON_MIN, LON_MAX, N_COLS)
    lat = np.linspace(LAT_MAX, LAT_MIN, N_ROWS)  # north to south

    # ---- Subset indices for BRB ----
    Ix = np.where((lon >= BRB_LON[0]) & (lon <= BRB_LON[1]))[0]
    Iy = np.where((lat >= BRB_LAT[0]) & (lat <= BRB_LAT[1]))[0]
    lon_sub = lon[Ix]
    lat_sub = lat[Iy]
    ny, nx = len(Iy), len(Ix)
    expected_shape = (ny, nx)

    print(f'BRB subset: {ny} x {nx} pixels')

    # ---- Build date list ----
    all_dates = build_date_list(WY)
    n_days = len(all_dates)
    date_strs = [d.strftime('%Y-%m-%d') for d in all_dates]

    print(f'Water Year {WY}: {date_strs[0]} to {date_strs[-1]} ({n_days} days)')

    # ---- Initialize output arrays ----
    snodas = {
        'WY':      WY,
        'lat':     lat_sub,
        'lon':     lon_sub,
        'dates':   np.array([d.toordinal() for d in all_dates]),  # Python ordinal
        'datestr': np.array(date_strs),
    }
    for name in PROD_NAMES:
        snodas[name] = np.full((ny, nx, n_days), np.nan, dtype=np.float32)

    # ---- Main download loop ----
    print(f'\n=== Downloading SNODAS data for WY{WY} ===')

    for d, (dt, ds) in enumerate(zip(all_dates, date_strs)):
        yyyy, mm, dd_day = dt.year, dt.month, dt.day
        month_dir = f'{mm:02d}_{MONTH_NAMES[mm - 1]}'
        tar_name  = f'SNODAS_{yyyy:04d}{mm:02d}{dd_day:02d}.tar'
        npz_file  = OUT_DIR / f'SNODAS_BRB_{yyyy:04d}{mm:02d}{dd_day:02d}.npz'

        # ---- Check cache ----
        cached = load_day_cache(npz_file, expected_shape)
        if cached is not None:
            for name in PROD_NAMES:
                if name in cached:
                    snodas[name][:, :, d] = cached[name]
            if (d + 1) % 30 == 0:
                print(f'  [{d+1}/{n_days}] {ds} - loaded from cache')
            continue

        # ---- Build URL and download ----
        url      = f'{BASE_URL}{yyyy:04d}/{month_dir}/{tar_name}'
        tar_path = TEMP_DIR / tar_name

        print(f'  [{d+1}/{n_days}] {ds} ... ', end='', flush=True)

        if not download_tar(url, tar_path):
            continue

        if not extract_tar(tar_path, TEMP_DIR):
            continue

        # ---- Process each variable ----
        day_data: dict = {}
        for code, name in zip(PROD_CODES, PROD_NAMES):
            # Find the .dat.gz file matching this product code
            gz_files = sorted(TEMP_DIR.glob(f'*{code}*.dat.gz'))
            if not gz_files:
                continue

            # Decompress
            dat_path = gunzip_file(gz_files[0])

            # Find the .dat file (gunzip may produce a different name)
            dat_files = sorted(TEMP_DIR.glob(f'*{code}*.dat'))
            if not dat_files:
                continue

            # Read full CONUS grid, subset to BRB
            D = read_snodas_binary(dat_files[0])
            D_sub = D[np.ix_(Iy, Ix)].astype(np.float32)

            snodas[name][:, :, d] = D_sub
            day_data[name] = D_sub

            dat_files[0].unlink()

        # ---- Clean up remaining temp files ----
        clean_temp_dir(TEMP_DIR)

        # ---- Save daily cache ----
        day_data['lat'] = lat_sub
        day_data['lon'] = lon_sub
        day_data['date'] = np.array([dt.toordinal()])
        np.savez(npz_file, **day_data)

        print('OK')

    # ---- Replace fill value -9999 with NaN ----
    print('\nCleaning no-data values...')
    for name in PROD_NAMES:
        snodas[name][snodas[name] == -9999] = np.nan

    # ---- Unit conversions ----
    print('Converting units...')
    apply_unit_conversions(snodas)

    # ---- Save complete water year ----
    out_file = OUT_DIR / f'SNODAS_BRB_WY{WY}.npz'
    np.savez(out_file, **snodas)
    print(f'\nSaved: {out_file}')

    # ---- Summary ----
    print(f'\n=== Download Summary ===')
    print(f'Water Year: {WY}')
    print(f'Region: Boise River Basin ({ny} x {nx} pixels)')
    print('Variables:')
    for name in PROD_NAMES:
        arr = snodas[name]
        n_valid = int(np.sum(~np.isnan(arr)))
        n_total = arr.size
        print(f'  {name}: {100 * n_valid / n_total:.1f}% valid data')

    # Quick peek: April 1 SWE
    apr1 = date(WY, 4, 1)
    apr1_idx = min(range(n_days), key=lambda i: abs((all_dates[i] - apr1).days))
    apr1_swe   = snodas['SWE'][:, :, apr1_idx]
    apr1_depth = snodas['Depth'][:, :, apr1_idx]

    print('\nApril 1 basin statistics:')
    print(f'  Mean SWE:   {np.nanmean(apr1_swe) * 100:.1f} cm')
    print(f'  Max SWE:    {np.nanmax(apr1_swe) * 100:.1f} cm')
    print(f'  Mean Depth: {np.nanmean(apr1_depth) * 100:.1f} cm')
    print(f'  Max Depth:  {np.nanmax(apr1_depth) * 100:.1f} cm')

    size_mb = out_file.stat().st_size / 1e6
    print(f'\nOutput file: {out_file} ({size_mb:.1f} MB)')
    print('\nDone! Run plotSNODAS_BRB.py to visualize.')


if __name__ == '__main__':
    main()
